### PR TITLE
Feature update

### PR DESCRIPTION
**Pull request by terminator2k2**
add Super FX Fast Rom
add BSX Loading

### CODE REVIEW

**Comment by agg23** [comment]
Please split the GSU change into a separate PR

File: src/fpga/core/rtl/chip/BSX/bsx121-124.mif [hunk 1]
```diff
@@ -28,20 +28,20 @@ CONTENT BEGIN
...
 	014  :   6F;
 	015  :   20;
-	016  :   4D;
-	017  :   69;
-	018  :   53;
-	019  :   54;
-	01A  :   65;
-	01B  :   72;
-	01C  :   21;
+	016  :   41;
```
Comment (agg23) [inline src/fpga/core/rtl/chip/BSX/bsx121-124.mif:31]: Add a comment (you can do that in mifs) indicating that this is the source text displayed in the games, and write out the text that this represents.

File: src/support/check_header.asm [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 constant output_address = 0x1904
 
-constant rambuf = 0x1b00 // Starts at 0xZZZBD in ROM
-constant header_start_mem = rambuf + 0x3 // Offset to 0xZZZC0
+constant rambuf = 0x1b00 // Starts at 0xZZZB2 in ROM
+constant header_start_mem = rambuf + 14 // Offset to 0xZZZC0
 
 // Address into header/ROM
-constant gsu_ramz_addr = rambuf // FBC
+constant gsu_ramz_addr = rambuf + 11 // FBD
```
Comment (agg23) [inline src/support/check_header.asm:8]: Good catch

File: src/support/check_header.asm [hunk 3]
```diff
@@ -50,7 +50,9 @@ macro check_header(variable base_address_input, variable output_address_input) {
   call validate_simple_values
   call choose_ramsz
   call choose_chip_type
+  call check_bsx
```
Comment (agg23) [inline src/support/check_header.asm:53]: This should be called at the end of `choose_chip_type`, not here.

File: src/support/check_header.asm [hunk 4]
```diff
@@ -50,7 +50,9 @@ macro check_header(variable base_address_input, variable output_address_input) {
...
   call choose_region
+
```
Comment (agg23) [inline src/support/check_header.asm:55]: Please remove unnecessary newlines

File: src/support/check_header.asm [hunk 5]
```diff
@@ -398,11 +402,53 @@ choose_region:
...
+  check_value_equality(rambuf + 1, 0x5A) // 'Z'
+  jp c, condition_three // Carry or Zero means <=
+  jp z, condition_three
+
+  condition_two:
+  check_value_equality(rambuf + 1, 0x30) // '0'
+  jp c, not_bsx
+  check_value_equality(rambuf + 1, 0x39) // '9'
+  jp c, condition_three // Carry or Zero means <=
+  jp z, condition_three
```
Comment (agg23) [inline src/support/check_header.asm:427]: Should jump to `not_bsx` if this falls through

File: src/support/check_header.asm [hunk 6]
```diff
@@ -398,11 +402,53 @@ choose_region:
...
+  jp z, is_bsx
+
+  check_value_equality(rambuf + 4, 0)
+  jp nz, not_bsx
+  check_value_equality(rambuf + 10, 0)
+  jp nz, not_bsx
+
+  is_bsx:
+  // Set chip_type lower nibble to 3
+  or r12,#0x30
```
Comment (agg23) [inline src/support/check_header.asm:441]: That is not the lower nibble

Comment (agg23) [reply src/support/check_header.asm:441]: This doesn't appear to match the MiSTer logic. If the above logic passes, then the ROM type is denoted as 3, which is the low nibble of the chip type value.

The upper nibble is only set if the file is detected to be a BSX BIOS, which is not checked for here.

File: src/support/loader.asm [hunk 7]
```diff
@@ -157,7 +158,7 @@ bit_bsx:
 cmp r4,#0x30 // Check if BSX
 jp nz, check_pal
 log_string("Using BSX")
-// It's BSX
+jp expansion_core// It's BSX
```
Comment (agg23) [inline src/support/loader.asm:161]: This doesn't do anything

**Review by agg23 — COMMENTED** [review]
It generally looks fine. It will be a while before I can test it probably, and I need you to separate out the GSU changes.

File: src/support/loader.asm (hunk 7 above)
Comment (terminator2k2) [reply src/support/loader.asm:161]: hey agg, can you refuse / cancel the whole pull request ? ive tried from my end but cant seem to find an option to delete/cancel. I have changed a few things.

also i dont think we will ever get the pocket core to load bsx roms like mister does as mister detects bs roms via the file extension ".bs"
https://github.com/MiSTer-devel/Main_MiSTer/blob/222d61776f3cc534f1e48ec1a6e31a3bc1588752/user_io.cpp#L2506

It also detects the bios which isnt a problem(we can do that)  also the header is modified on load on some roms as they were playable only at set times. So the way i load bsx is detect the bsx bios header which concatenated with each rom

**Pull request closed without merging**